service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/RequestStatementMapper.java

```diff
@@ -41,182 +42,238 @@
     private static final String SPECIALTY = "Specialty: ";
-    private static final String REASON_CODE = "Reason Codes: ";
+    private static final String REASON_CODES = "Reason Codes: ";
+
     private static final String DEFAULT_REASON_CODE_XML = "<code code=\"3457005\" displayName=\"Patient referral\" codeSystem=\"2.16.840.1"
         + ".113883.2.1.3.2.4.15\"/>";
     private static final String NOTE = "Annotation: %s @ %s %s";
     private static final String COMMA = ", ";
 
     private final MessageContext messageContext;
     private final CodeableConceptCdMapper codeableConceptCdMapper;
+    private final ParticipantMapper participantMapper;
 
     public String mapReferralRequestToRequestStatement(ReferralRequest referralRequest, boolean isNested) {
-        var requestStatementTemplateParameters = RequestStatementTemplateParameters.builder()
-            .requestStatementId(messageContext.getIdMapper().getOrNew(ResourceType.ReferralRequest, referralRequest.getId()))
-            .isNested(isNested)
-            .availabilityTime(StatementTimeMappingUtils.prepareAvailabilityTimeForReferralRequest(referralRequest))
-            .description(buildDescription(referralRequest))
-            .code(buildCode(referralRequest))
-            .build();
-
-        if (!referralRequest.hasReasonCode()) {
-            requestStatementTemplateParameters.setDefaultReasonCode(DEFAULT_REASON_CODE_XML);
+        return new InnerMapper(referralRequest, isNested).map();
+    }
+
+    @RequiredArgsConstructor
+    private class InnerMapper {
+        private final ReferralRequest referralRequest;
+        private final boolean isNested;
+        private final RequestStatementTemplateParametersBuilder templateParameters = RequestStatementTemplateParameters.builder();
+        private String text = StringUtils.EMPTY;
+
+        private String map() {
+            final IdMapper idMapper = messageContext.getIdMapper();
+            templateParameters
+                .requestStatementId(idMapper.getOrNew(ResourceType.ReferralRequest, referralRequest.getId()))
+                .isNested(isNested)
+                .availabilityTime(StatementTimeMappingUtils.prepareAvailabilityTimeForReferralRequest(referralRequest))
+                .code(buildCode());
+
+            text = buildTextDescription();
+            prependDescriptionInfo();
+
+            if (!referralRequest.hasReasonCode()) {
+                templateParameters.defaultReasonCode(DEFAULT_REASON_CODE_XML);
+            }
+
+            if (referralRequest.hasRequester()
+                    && referralRequest.getRequester().hasAgent()
+                    && referralRequest.getRequester().getAgent().hasReference()) {
+                var requester = referralRequest.getRequester();
+                Reference agentRef = requester.getAgent();
+                var onBehalfOf = requester.hasOnBehalfOf() ? requester.getOnBehalfOf() : null;
+                processAgent(agentRef, onBehalfOf);
```

Comment:
what do you think about changing the method processAgent into 2 : getAgent and getText ?  I think process agent method has too much responsibilities at the moment.

Reply:
Agreed. The main difficulty was in maintaining that exception thrown for an unexpected agent type; please let me know what you think